// @module-federation/vite/lib/packageUtils-Bbc6r6__.mjs
import { createRequire } from "node:module";
import { existsSync, readFileSync, readdirSync } from "fs";
import { createRequire as createRequire$1 } from "module";
import path from "pathe";
//#region \0rolldown/runtime.js
var __require = /* @__PURE__ */ createRequire(import.meta.url);
//#endregion
//#region src/utils/logger.ts
const MODULE_FEDERATION_LOG_PREFIX = "[Module Federation]";
function formatModuleFederationMessage(message) {
	return `${MODULE_FEDERATION_LOG_PREFIX} ${message}`;
}
function createModuleFederationError(message) {
	return new Error(formatModuleFederationMessage(message));
}
function toConsoleArgs(message, rest = []) {
	if (typeof message === "string") return [formatModuleFederationMessage(message), ...rest];
	if (message === void 0) return [MODULE_FEDERATION_LOG_PREFIX, ...rest];
	return [
		MODULE_FEDERATION_LOG_PREFIX,
		message,
		...rest
	];
}
const moduleFederationConsole = {
	log(message, ...rest) {
		console.log(...toConsoleArgs(message, rest));
	},
	warn(message, ...rest) {
		console.warn(...toConsoleArgs(message, rest));
	},
	error(message, ...rest) {
		console.error(...toConsoleArgs(message, rest));
	}
};
moduleFederationConsole.log;
const mfWarn = moduleFederationConsole.warn;
const mfError = moduleFederationConsole.error;
//#endregion
//#region src/utils/packageUtils.ts
const dependencyPresenceCache = /* @__PURE__ */ new Map();
let packageDetectionCwd;
function getDependencyCacheKey(cwd, dependencyName) {
	return `${cwd}:${dependencyName}`;
}
function setPackageDetectionCwd(cwd) {
	packageDetectionCwd = cwd;
}
function getPackageDetectionCwd() {
	return packageDetectionCwd || process.cwd();
}
const DEFAULT_EXPORT_CONDITIONS = [
	"browser",
	"import",
	"module",
	"default",
	"require"
];
function resolveExportsEntry(exportsField, conditions = DEFAULT_EXPORT_CONDITIONS) {
	if (typeof exportsField === "string") return exportsField;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const rootExport = record["."];
	if (rootExport) return resolveExportsEntry(rootExport);
	for (const condition of conditions) {
		const target = resolveExportsEntry(record[condition], conditions);
		if (target) return target;
	}
	for (const target of Object.values(record)) {
		const resolved = resolveExportsEntry(target, conditions);
		if (resolved) return resolved;
	}
}
function getPackageExportsTarget(pkg, packageName, exportsField) {
	if (typeof exportsField === "string") return pkg === packageName ? exportsField : void 0;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const subpath = pkg === packageName ? "." : `.${pkg.slice(packageName.length)}`;
	if (subpath !== ".") return record[subpath];
	return record["."] ?? (!Object.keys(record).some((key) => key.startsWith(".")) ? record : void 0);
}
/**
* Escaping rules:
* Convert using the format __${mapping}__, where _ and $ are not allowed in npm package names but can be used in variable names.
*  @ => 1
*  / => 2
*  - => 3
*  . => 4
*/
/**
* Encodes a package name into a valid file name.
* @param {string} name - The package name, e.g., "@scope/xx-xx.xx".
* @returns {string} - The encoded file name.
*/
function packageNameEncode(name) {
	if (typeof name !== "string") throw createModuleFederationError("A string package name is required");
	return name.replace(/@/g, "_mf_0_").replace(/\//g, "_mf_1_").replace(/-/g, "_mf_2_").replace(/\./g, "_mf_3_");
}
/**
* Decodes an encoded file name back to the original package name.
* @param {string} encoded - The encoded file name, e.g., "_mf_0_scope_mf_1_xx_mf_2_xx_mf_3_xx".
* @returns {string} - The decoded package name.
*/
function packageNameDecode(encoded) {
	if (typeof encoded !== "string") throw createModuleFederationError("A string encoded file name is required");
	return encoded.replace(/_mf_0_/g, "@").replace(/_mf_1_/g, "/").replace(/_mf_2_/g, "-").replace(/_mf_3_/g, ".");
}
/**
* Removes any subpath from an npm package specifier and returns the package name only.
* @param {string} packageString - The package specifier, e.g., "@scope/pkg/runtime" or "react/jsx-runtime".
* @returns {string} - The base npm package name.
*/
function getPackageName(packageString) {
	const match = packageString.match(/^(?:@[^/]+\/)?[^/]+/);
	return match ? match[0] : packageString;
}
function getPackageNameFromNodeModulePath(source) {
	const normalized = source.replace(/\\/g, "/");
	const nodeModulesIndex = normalized.lastIndexOf("/node_modules/");
	if (nodeModulesIndex < 0) return;
	const parts = normalized.slice(nodeModulesIndex + 14).split("/");
	if (!parts[0]) return;
	if (parts[0].startsWith("@")) return parts[1] ? `${parts[0]}/${parts[1]}` : void 0;
	return parts[0];
}
function getSharedCacheKey(pkg, shareItem) {
	return shareItem.shareConfig.singleton || !shareItem.version ? pkg : `${pkg}@${shareItem.version}`;
}
function getInstalledPackageJson(pkg, opts) {
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	const tryReadPackageJson = (packageJsonPath) => {
		if (!existsSync(packageJsonPath)) return void 0;
		try {
			return {
				path: packageJsonPath,
				dir: path.dirname(packageJsonPath),
				packageJson: JSON.parse(readFileSync(packageJsonPath, "utf-8"))
			};
		} catch {
			return;
		}
	};
	const findPackageInPnpmStore = (startDir) => {
		let currentDir = startDir;
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const pnpmStoreDir = path.join(currentDir, "node_modules", ".pnpm");
			if (existsSync(pnpmStoreDir)) try {
				for (const entry of readdirSync(pnpmStoreDir, { withFileTypes: true })) {
					if (!entry.isDirectory()) continue;
					const candidate = tryReadPackageJson(path.join(pnpmStoreDir, entry.name, "node_modules", packageName, "package.json"));
					if (candidate?.packageJson.name === packageName) return candidate;
				}
			} catch {}
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
	};
	try {
		const projectRequire = createRequire$1(new URL(`file://${path.join(cwd, "package.json")}`));
		let resolvedPath;
		if (opts?.fromResolvedEntry) resolvedPath = opts.fromResolvedEntry;
		else try {
			resolvedPath = projectRequire.resolve(pkg);
		} catch {
			resolvedPath = projectRequire.resolve(packageName);
		}
		let currentDir = path.dirname(resolvedPath);
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const packageJsonPath = path.join(currentDir, "package.json");
			if (existsSync(packageJsonPath)) {
				const packageJsonContent = readFileSync(packageJsonPath, "utf-8");
				try {
					const packageJson = JSON.parse(packageJsonContent);
					if (packageJson.name === packageName) return {
						path: packageJsonPath,
						dir: currentDir,
						packageJson
					};
				} catch (error) {
					if (!(error instanceof SyntaxError)) throw error;
				}
			}
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
	} catch {
		let currentDir = cwd;
		const rootDir = path.parse(currentDir).root;
		while (true) {
			const directCandidate = tryReadPackageJson(path.join(currentDir, "node_modules", packageName, "package.json"));
			if (directCandidate?.packageJson.name === packageName) return directCandidate;
			if (currentDir === rootDir) break;
			currentDir = path.dirname(currentDir);
		}
		return findPackageInPnpmStore(cwd);
	}
}
function getInstalledPackageEntry(pkg, opts) {
	const installed = getInstalledPackageJson(pkg, opts);
	if (!installed) return void 0;
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	if (pkg !== packageName && opts?.resolveSubpathWithRequire !== false) try {
		return createRequire$1(new URL(`file://${path.join(cwd, "package.json")}`)).resolve(pkg);
	} catch {}
	const packageJson = installed.packageJson;
	const explicitEntry = resolveExportsEntry(getPackageExportsTarget(pkg, packageName, packageJson.exports), opts?.conditions) || (typeof packageJson.module === "string" ? packageJson.module : void 0) || (typeof packageJson.main === "string" ? packageJson.main : void 0) || "index.js";
	return path.join(installed.dir, explicitEntry);
}
/**
* Detect whether the current runtime is Vite 8+ by checking for a Vite version flag
* on the plugin hook context, with Rolldown metadata kept as a compatibility fallback.
*/
function getIsRolldown(ctx) {
	const viteVersion = ctx?.meta?.viteVersion;
	const viteMajor = Number(String(viteVersion ?? "").split(".")[0]);
	return Number.isFinite(viteMajor) && viteMajor >= 8 || !!ctx?.meta?.rolldownVersion;
}
/** Walk up from Vite `config.root` (Nuxt may point at `.nuxt` cache dirs). */
function isNuxtProjectRoot(root) {
	let dir = root;
	for (let i = 0; i < 8; i++) {
		if (hasPackageDependency("nuxt", dir) || hasPackageDependency("nuxt-nightly", dir)) return true;
		const parent = path.dirname(dir);
		if (parent === dir) break;
		dir = parent;
	}
	return false;
}
function hasPackageDependency(dependencyName, cwd = packageDetectionCwd || process.cwd()) {
	const cacheKey = getDependencyCacheKey(cwd, dependencyName);
	const cached = dependencyPresenceCache.get(cacheKey);
	if (cached !== void 0) return cached;
	try {
		const packageJson = JSON.parse(readFileSync(path.join(cwd, "package.json"), "utf8"));
		const hasDependency = [
			packageJson.dependencies,
			packageJson.devDependencies,
			packageJson.peerDependencies,
			packageJson.optionalDependencies
		].some((deps) => !!deps?.[dependencyName]);
		dependencyPresenceCache.set(cacheKey, hasDependency);
		return hasDependency;
	} catch {
		dependencyPresenceCache.set(cacheKey, false);
		return false;
	}
}
//#endregion
export { getPackageName as a, hasPackageDependency as c, packageNameEncode as d, setPackageDetectionCwd as f, __require as g, mfWarn as h, getPackageDetectionCwd as i, isNuxtProjectRoot as l, mfError as m, getInstalledPackageJson as n, getPackageNameFromNodeModulePath as o, createModuleFederationError as p, getIsRolldown as r, getSharedCacheKey as s, getInstalledPackageEntry as t, packageNameDecode as u };
