// @module-federation/vite/lib/packageUtils-se-UDhCa.cjs
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let fs = require("fs");
let module$1 = require("module");
let pathe = require("pathe");
pathe = __toESM(pathe);
//#region src/utils/logger.ts
const MODULE_FEDERATION_LOG_PREFIX = "[Module Federation]";
function formatModuleFederationMessage(message) {
	return `${MODULE_FEDERATION_LOG_PREFIX} ${message}`;
}
function createModuleFederationError(message) {
	return new Error(formatModuleFederationMessage(message));
}
function toConsoleArgs(message, rest = []) {
	if (typeof message === "string") return [formatModuleFederationMessage(message), ...rest];
	if (message === void 0) return [MODULE_FEDERATION_LOG_PREFIX, ...rest];
	return [
		MODULE_FEDERATION_LOG_PREFIX,
		message,
		...rest
	];
}
const moduleFederationConsole = {
	log(message, ...rest) {
		console.log(...toConsoleArgs(message, rest));
	},
	warn(message, ...rest) {
		console.warn(...toConsoleArgs(message, rest));
	},
	error(message, ...rest) {
		console.error(...toConsoleArgs(message, rest));
	}
};
moduleFederationConsole.log;
const mfWarn = moduleFederationConsole.warn;
const mfError = moduleFederationConsole.error;
//#endregion
//#region src/utils/packageUtils.ts
const dependencyPresenceCache = /* @__PURE__ */ new Map();
let packageDetectionCwd;
function getDependencyCacheKey(cwd, dependencyName) {
	return `${cwd}:${dependencyName}`;
}
function setPackageDetectionCwd(cwd) {
	packageDetectionCwd = cwd;
}
function getPackageDetectionCwd() {
	return packageDetectionCwd || process.cwd();
}
const DEFAULT_EXPORT_CONDITIONS = [
	"browser",
	"import",
	"module",
	"default",
	"require"
];
function resolveExportsEntry(exportsField, conditions = DEFAULT_EXPORT_CONDITIONS) {
	if (typeof exportsField === "string") return exportsField;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const rootExport = record["."];
	if (rootExport) return resolveExportsEntry(rootExport);
	for (const condition of conditions) {
		const target = resolveExportsEntry(record[condition], conditions);
		if (target) return target;
	}
	for (const target of Object.values(record)) {
		const resolved = resolveExportsEntry(target, conditions);
		if (resolved) return resolved;
	}
}
function getPackageExportsTarget(pkg, packageName, exportsField) {
	if (typeof exportsField === "string") return pkg === packageName ? exportsField : void 0;
	if (!exportsField || typeof exportsField !== "object") return void 0;
	const record = exportsField;
	const subpath = pkg === packageName ? "." : `.${pkg.slice(packageName.length)}`;
	if (subpath !== ".") return record[subpath];
	return record["."] ?? (!Object.keys(record).some((key) => key.startsWith(".")) ? record : void 0);
}
/**
* Escaping rules:
* Convert using the format __${mapping}__, where _ and $ are not allowed in npm package names but can be used in variable names.
*  @ => 1
*  / => 2
*  - => 3
*  . => 4
*/
/**
* Encodes a package name into a valid file name.
* @param {string} name - The package name, e.g., "@scope/xx-xx.xx".
* @returns {string} - The encoded file name.
*/
function packageNameEncode(name) {
	if (typeof name !== "string") throw createModuleFederationError("A string package name is required");
	return name.replace(/@/g, "_mf_0_").replace(/\//g, "_mf_1_").replace(/-/g, "_mf_2_").replace(/\./g, "_mf_3_");
}
/**
* Decodes an encoded file name back to the original package name.
* @param {string} encoded - The encoded file name, e.g., "_mf_0_scope_mf_1_xx_mf_2_xx_mf_3_xx".
* @returns {string} - The decoded package name.
*/
function packageNameDecode(encoded) {
	if (typeof encoded !== "string") throw createModuleFederationError("A string encoded file name is required");
	return encoded.replace(/_mf_0_/g, "@").replace(/_mf_1_/g, "/").replace(/_mf_2_/g, "-").replace(/_mf_3_/g, ".");
}
/**
* Removes any subpath from an npm package specifier and returns the package name only.
* @param {string} packageString - The package specifier, e.g., "@scope/pkg/runtime" or "react/jsx-runtime".
* @returns {string} - The base npm package name.
*/
function getPackageName(packageString) {
	const match = packageString.match(/^(?:@[^/]+\/)?[^/]+/);
	return match ? match[0] : packageString;
}
function getPackageNameFromNodeModulePath(source) {
	const normalized = source.replace(/\\/g, "/");
	const nodeModulesIndex = normalized.lastIndexOf("/node_modules/");
	if (nodeModulesIndex < 0) return;
	const parts = normalized.slice(nodeModulesIndex + 14).split("/");
	if (!parts[0]) return;
	if (parts[0].startsWith("@")) return parts[1] ? `${parts[0]}/${parts[1]}` : void 0;
	return parts[0];
}
function getSharedCacheKey(pkg, shareItem) {
	return shareItem.shareConfig.singleton || !shareItem.version ? pkg : `${pkg}@${shareItem.version}`;
}
function getInstalledPackageJson(pkg, opts) {
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	const tryReadPackageJson = (packageJsonPath) => {
		if (!(0, fs.existsSync)(packageJsonPath)) return void 0;
		try {
			return {
				path: packageJsonPath,
				dir: pathe.default.dirname(packageJsonPath),
				packageJson: JSON.parse((0, fs.readFileSync)(packageJsonPath, "utf-8"))
			};
		} catch {
			return;
		}
	};
	const findPackageInPnpmStore = (startDir) => {
		let currentDir = startDir;
		const rootDir = pathe.default.parse(currentDir).root;
		while (true) {
			const pnpmStoreDir = pathe.default.join(currentDir, "node_modules", ".pnpm");
			if ((0, fs.existsSync)(pnpmStoreDir)) try {
				for (const entry of (0, fs.readdirSync)(pnpmStoreDir, { withFileTypes: true })) {
					if (!entry.isDirectory()) continue;
					const candidate = tryReadPackageJson(pathe.default.join(pnpmStoreDir, entry.name, "node_modules", packageName, "package.json"));
					if (candidate?.packageJson.name === packageName) return candidate;
				}
			} catch {}
			if (currentDir === rootDir) break;
			currentDir = pathe.default.dirname(currentDir);
		}
	};
	try {
		const projectRequire = (0, module$1.createRequire)(new URL(`file://${pathe.default.join(cwd, "package.json")}`));
		let resolvedPath;
		if (opts?.fromResolvedEntry) resolvedPath = opts.fromResolvedEntry;
		else try {
			resolvedPath = projectRequire.resolve(pkg);
		} catch {
			resolvedPath = projectRequire.resolve(packageName);
		}
		let currentDir = pathe.default.dirname(resolvedPath);
		const rootDir = pathe.default.parse(currentDir).root;
		while (true) {
			const packageJsonPath = pathe.default.join(currentDir, "package.json");
			if ((0, fs.existsSync)(packageJsonPath)) {
				const packageJsonContent = (0, fs.readFileSync)(packageJsonPath, "utf-8");
				try {
					const packageJson = JSON.parse(packageJsonContent);
					if (packageJson.name === packageName) return {
						path: packageJsonPath,
						dir: currentDir,
						packageJson
					};
				} catch (error) {
					if (!(error instanceof SyntaxError)) throw error;
				}
			}
			if (currentDir === rootDir) break;
			currentDir = pathe.default.dirname(currentDir);
		}
	} catch {
		let currentDir = cwd;
		const rootDir = pathe.default.parse(currentDir).root;
		while (true) {
			const directCandidate = tryReadPackageJson(pathe.default.join(currentDir, "node_modules", packageName, "package.json"));
			if (directCandidate?.packageJson.name === packageName) return directCandidate;
			if (currentDir === rootDir) break;
			currentDir = pathe.default.dirname(currentDir);
		}
		return findPackageInPnpmStore(cwd);
	}
}
function getInstalledPackageEntry(pkg, opts) {
	const installed = getInstalledPackageJson(pkg, opts);
	if (!installed) return void 0;
	const cwd = opts?.cwd || getPackageDetectionCwd();
	const packageName = opts?.packageName || getPackageName(pkg);
	if (pkg !== packageName && opts?.resolveSubpathWithRequire !== false) try {
		return (0, module$1.createRequire)(new URL(`file://${pathe.default.join(cwd, "package.json")}`)).resolve(pkg);
	} catch {}
	const packageJson = installed.packageJson;
	const explicitEntry = resolveExportsEntry(getPackageExportsTarget(pkg, packageName, packageJson.exports), opts?.conditions) || (typeof packageJson.module === "string" ? packageJson.module : void 0) || (typeof packageJson.main === "string" ? packageJson.main : void 0) || "index.js";
	return pathe.default.join(installed.dir, explicitEntry);
}
/**
* Detect whether the current runtime is Vite 8+ by checking for a Vite version flag
* on the plugin hook context, with Rolldown metadata kept as a compatibility fallback.
*/
function getIsRolldown(ctx) {
	const viteVersion = ctx?.meta?.viteVersion;
	const viteMajor = Number(String(viteVersion ?? "").split(".")[0]);
	return Number.isFinite(viteMajor) && viteMajor >= 8 || !!ctx?.meta?.rolldownVersion;
}
/** Walk up from Vite `config.root` (Nuxt may point at `.nuxt` cache dirs). */
function isNuxtProjectRoot(root) {
	let dir = root;
	for (let i = 0; i < 8; i++) {
		if (hasPackageDependency("nuxt", dir) || hasPackageDependency("nuxt-nightly", dir)) return true;
		const parent = pathe.default.dirname(dir);
		if (parent === dir) break;
		dir = parent;
	}
	return false;
}
function hasPackageDependency(dependencyName, cwd = packageDetectionCwd || process.cwd()) {
	const cacheKey = getDependencyCacheKey(cwd, dependencyName);
	const cached = dependencyPresenceCache.get(cacheKey);
	if (cached !== void 0) return cached;
	try {
		const packageJson = JSON.parse((0, fs.readFileSync)(pathe.default.join(cwd, "package.json"), "utf8"));
		const hasDependency = [
			packageJson.dependencies,
			packageJson.devDependencies,
			packageJson.peerDependencies,
			packageJson.optionalDependencies
		].some((deps) => !!deps?.[dependencyName]);
		dependencyPresenceCache.set(cacheKey, hasDependency);
		return hasDependency;
	} catch {
		dependencyPresenceCache.set(cacheKey, false);
		return false;
	}
}
//#endregion
Object.defineProperty(exports, "__toESM", {
	enumerable: true,
	get: function() {
		return __toESM;
	}
});
Object.defineProperty(exports, "createModuleFederationError", {
	enumerable: true,
	get: function() {
		return createModuleFederationError;
	}
});
Object.defineProperty(exports, "getInstalledPackageEntry", {
	enumerable: true,
	get: function() {
		return getInstalledPackageEntry;
	}
});
Object.defineProperty(exports, "getInstalledPackageJson", {
	enumerable: true,
	get: function() {
		return getInstalledPackageJson;
	}
});
Object.defineProperty(exports, "getIsRolldown", {
	enumerable: true,
	get: function() {
		return getIsRolldown;
	}
});
Object.defineProperty(exports, "getPackageDetectionCwd", {
	enumerable: true,
	get: function() {
		return getPackageDetectionCwd;
	}
});
Object.defineProperty(exports, "getPackageName", {
	enumerable: true,
	get: function() {
		return getPackageName;
	}
});
Object.defineProperty(exports, "getPackageNameFromNodeModulePath", {
	enumerable: true,
	get: function() {
		return getPackageNameFromNodeModulePath;
	}
});
Object.defineProperty(exports, "getSharedCacheKey", {
	enumerable: true,
	get: function() {
		return getSharedCacheKey;
	}
});
Object.defineProperty(exports, "hasPackageDependency", {
	enumerable: true,
	get: function() {
		return hasPackageDependency;
	}
});
Object.defineProperty(exports, "isNuxtProjectRoot", {
	enumerable: true,
	get: function() {
		return isNuxtProjectRoot;
	}
});
Object.defineProperty(exports, "mfError", {
	enumerable: true,
	get: function() {
		return mfError;
	}
});
Object.defineProperty(exports, "mfWarn", {
	enumerable: true,
	get: function() {
		return mfWarn;
	}
});
Object.defineProperty(exports, "packageNameDecode", {
	enumerable: true,
	get: function() {
		return packageNameDecode;
	}
});
Object.defineProperty(exports, "packageNameEncode", {
	enumerable: true,
	get: function() {
		return packageNameEncode;
	}
});
Object.defineProperty(exports, "setPackageDetectionCwd", {
	enumerable: true,
	get: function() {
		return setPackageDetectionCwd;
	}
});
